src/omv/imlib/binary.c

```diff
@@ -260,54 +260,84 @@ static void imlib_b_and_line_op(image_t *img, int line, void *other, void *data,
 
     switch (img->pixfmt) {
         case PIXFORMAT_BINARY: {
-            uint32_t *data = IMAGE_COMPUTE_BINARY_PIXEL_ROW_PTR(img, line);
+            uint32_t *row = IMAGE_COMPUTE_BINARY_PIXEL_ROW_PTR(img, line);
 
             if (!mask) {
-                for (int i = 0, j = IMAGE_BINARY_LINE_LEN(img); i < j; i++) {
-                    data[i] &= ((uint32_t *) other)[i];
+                for (int x = 0; x < img->w; x++) {
+                    int dataPixel = IMAGE_GET_BINARY_PIXEL_FAST(row, x);
+                    int otherPixel = IMAGE_GET_BINARY_PIXEL_FAST(((uint32_t *) other), x);
+                    int p = dataPixel & otherPixel;
+                    IMAGE_PUT_BINARY_PIXEL_FAST(row, x, p);
                 }
             } else {
-                for (int i = 0, j = img->w; i < j; i++) {
-                    if (image_get_mask_pixel(mask, i, line)) {
-                        IMAGE_PUT_BINARY_PIXEL_FAST(data, i,
-                                                    (IMAGE_GET_BINARY_PIXEL_FAST(data, i)
-                                                     & IMAGE_GET_BINARY_PIXEL_FAST(((uint32_t *) other), i)));
+                for (int x = 0; x < img->w; x++) {
+                    if (image_get_mask_pixel(mask, x, line)) {
+                        int dataPixel = IMAGE_GET_BINARY_PIXEL_FAST(row, x);
+                        int otherPixel = IMAGE_GET_BINARY_PIXEL_FAST(((uint32_t *) other), x);
+                        int p = dataPixel & otherPixel;
+                        IMAGE_PUT_BINARY_PIXEL_FAST(row, x, p);
                     }
                 }
             }
             break;
         }
         case PIXFORMAT_GRAYSCALE: {
-            uint8_t *data = IMAGE_COMPUTE_GRAYSCALE_PIXEL_ROW_PTR(img, line);
+            uint8_t *row = IMAGE_COMPUTE_GRAYSCALE_PIXEL_ROW_PTR(img, line);
 
             if (!mask) {
-                for (int i = 0, j = IMAGE_GRAYSCALE_LINE_LEN(img); i < j; i++) {
-                    data[i] &= ((uint8_t *) other)[i];
+                int x = 0;
+                #if (__ARM_ARCH > 6)
+                for (; (img->w - x) >= 4; x += 4) {
+                    int dataPixels = *((uint32_t *) (row + x));
+                    int otherPixels = *((uint32_t *) (((uint8_t *) other) + x));
```

Comment:
Artifact from breaking out the PR from the final form. Easy to fix.

Reply:
This is needed actually to get the pointer math right because I'm computing the address based on x and then doing an unaligned load of 32-bits. This is why the code is wrapped by __ARM_ARCH > 6.